Basic LDAP support for domain authentication

Armored-Dragon opened the pull request:
This is a implementation for using LDAP services as an alternative authentication method to WordPress for domain servers.

Closes #905 as complete.
Closes #906 as complete.

General TODO:
- [x] Fix LoginDialog.qml style issues.
- [x] LDAP server groups as roles.
- [x] LDAPAccount.cpp (better) error handling.
- [x] LDAPAccount.cpp expose setting server url.
- [x] Fix LDAP accounts with capital letters can not sign in.
- [x] Documentation.
- [ ] Include LDAP library.


Out of scope TODO. (A follow up pull request will be opened to address these issues)
- Fix LoginDialog.qml not being updated on failed LDAP signin attempt
- Find a better way to store sensitive account information
- Automatically use the correct authentication protocol (LDAP vs WordPress(OAuth2)) without client interaction.
- https://github.com/overte-org/overte/issues/208 (This might be addressable directly, otherwise it may be possible to sneak a fix in)
- "No valid access token present" error.
- https://github.com/overte-org/overte/pull/1060

Armored-Dragon committed:
Initial Support.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Initial Support.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Initial Support.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Build fix for Linux and Windows.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Build fix for Linux.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Build fix for Linux and Windows.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Build fix for Linux.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Build fix for Linux and Windows.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added the vcpkg files for windows.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added the vcpkg files for windows.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Fixed position issues on "Sign up" link.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Fixed position issues on "Sign up" link.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Fixed position issues on "Sign up" link.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Domain usernames are now case-sensitive.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Domain usernames are now case-sensitive.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added persistence to authenticated domains though LDAP.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added persistence to authenticated domains though LDAP.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added persistence to authenticated domains though LDAP.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added roles from LDAP user.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added roles from LDAP user.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Added roles from LDAP user.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon commented:
Strange issue: I logged into the directory server and then attempted to connect to my localhost LDAP domain server. This caused some strange signature validation error which would be caught on [line 788](https://github.com/Armored-Dragon/overte/blob/4ad0d73bae05d3661ec02f9c0c0fef2cb1603c7a/domain-server/src/DomainGatekeeper.cpp#L788).
I rebuilt a few times after some changes and a revert to the point to when the issue started somehow fixed the signature validation?
I suspect there may be issues I have not considered from converting the domainUsername value to be case sensitive. Case insensitivity was initially pushed on commit [9d9c34a](https://github.com/overte-org/overte/pull/1159/commits/9d9c34a36d43e12c0e6ad0ec664ca8cec3f10c35) and fully implemented on commit [4ad0d73](https://github.com/overte-org/overte/pull/1159/commits/4ad0d73bae05d3661ec02f9c0c0fef2cb1603c7a).
I will need to look into this later to make sure I didn't break anything.

pull[bot] committed:
[pull] master from overte-org:master (#93)

* mirrors wip

* fix view + projection, texture flipping, billboarding

* wip portals

* wip

* fix cpu frustum culling (hacky?)

* fix mirrors in deferred

* mirrors on models + text

* portals use exit as ignoreItem

* cleanup

* entity tags

* wild guess to handle view correction, hide portalExitID in create when mirrorMode != portal

* let's try this??

* plz

* promising

* fix paramsOffset and view flipping

* portals shouldn't flip

* break when tag found

* fix portal view calculation

* Revert "Mirrors + Portals"

* Revert "Revert "Mirrors + Portals""

* web entity wantsKeyboardFocus property

* fix typo

* move audio zones to zone entity properties

* fix audio zones in create

* set dynamic factory

* new procecural particle entity type

* fix particle intersection

* shorten create labels

* fix 0 update props case

* Ability to smooth model animations

* sound entities

* fix layered simulate items

* fix stereo sound speed

* support non-localOnly sound avatar entities

* add sound url prompt

* support registration point, improve locking

* remove keyboardRasied

* locking attempt #2

* fix keyboardRasied typo

* add default particle props

* add unlit property for shapes

* Merge branch master into protocol_changes

* add ambient light color

* fix create issue

* fix create issue

* add tonemapping props to zones, wip ambient occlusion

* wip ambient occlusion

* it's working!

* remove attachments

* fix non-localOnly positional sounds not updating

* change AO default to HBAO, remove from create

* more graphics options

* fix AO setting + effects in mirrors

* fix AA in mirrors

* alezia's fixes

* fix haze in mirrors

* add comment for SKYBOX_DISTANCE

* new line

* model loading priority updates over time, takes into account out of bounds, avatar entities have higher priority, and fsts can specify to wait for wearables to load before rendering

* add loadPriority to model entities, working on other avatars waitForWearables

* fix build error

* try to fix isServer assert

* fix stats + waitForWearables

* Listen for click instead of release.

* Reverted initial commit. Implemented hack to listen for menu click events.

* Missed some reverts.

* Missed another one.

* Prevent duplicate actions.

* Added extra needed checks.

* Fix without formatting? (#91)

* Hopefully fixed formatting.

* Things can't be too easy.

* Remove google poly

* automated entity property serialization

* cleanup + automate EntityPropertyFlags

* text vertical alignment, use uint8_t for entity property enums, fix text recalculating too often

* fix text size

* Update interface/resources/controllers/keyboardMouse.json

Co-authored-by: HifiExperiments <53453710+HifiExperiments@users.noreply.github.com>

* fix component mode serialization

* Fixed mouse look in selfie mode.

* fix text debug assert on invalid or unloaded font

* missed some enums

* fix ADD_GROUP_PROPERTY_TO_MAP

* fix PROP_GRAB_EQUIPPABLE_INDICATOR_URL missing urlPermission

* fix KeyLightPropertyGroup legacy properties

* fix PolyLineEntityItem::getEntityProperties

* comment cmake script

* fix copyright

* Replaced key value with key text.
Added additional comment about the specific delete key used.

* weekly promoted place

Highlight the first place in the list as the weekly promoted place

* Fixed lingering references to `avatarIcon`.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

* Adding icon for "Grab And Equip" section

Adding icon for "Grab And Equip" section

* Add "Grab And Equip" section

Add "Grab And Equip" section for the grabbale and Equipable groups of properties.

* Add files via upload

* Add tooltips for the "Grab and Equip" properties

Add the tooltips for the "Grab and Equip" properties

* Text adjustments for grab.equippable

Text adjustments for grab.equippable

* Make Maturity Filter persisted

Make Maturity Filter persisted and with a default value (Teen & Everyone)

* Adjust the default value for maturity

Adjust the default value for maturity

* move "triggerable" under GRAB & EQUIP

move "triggerable" under GRAB & EQUIP

* Remove hifi-screenshare
Cherry picked and updated from Tivoli dd5b6ea6ee5597a06603e16509640e7ed18106bb

Co-authored-by: Julian Groß <julian.g@posteo.de>

* Insert placeholder to not break protocol yet.

* Fix incorrectly resolved merge conflict, left too much code.

* Fixes based on review comments on previous PR

* Remove code accidentally re-added during a conflict fix

* bump protocol

* rebuild fonts with full charset (NOT -allglyphs)

* Attempt at fixing Windows master branch builds

* Change minimum angular velocity to a lower one

* Fix Uuid.NULL behavior

---------

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>
Co-authored-by: HifiExperiments <thingsandstuffblog@gmail.com>
Co-authored-by: ksuprynowicz <ksuprynowicz@post.pl>
Co-authored-by: Dale Glass <51060919+daleglass@users.noreply.github.com>
Co-authored-by: HifiExperiments <53453710+HifiExperiments@users.noreply.github.com>
Co-authored-by: Julian Groß <julian.g@posteo.de>
Co-authored-by: armored-dragon <publicmail@armoreddragon.com>
Co-authored-by: Armored-Dragon <github56254@armoreddragon.com>
Co-authored-by: Alezia Kurdis <60075796+AleziaKurdis@users.noreply.github.com>
Co-authored-by: Maki <mxmcube@gmail.com>
Co-authored-by: Dale Glass <dale@daleglass.net>

Armored-Dragon committed:
Merge branch 'overte-org:master' into master

Armored-Dragon committed:
Read from the admin panel for LDAP settings.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Merge remote-tracking branch 'origin/feature/basicLDAP' into feature/basicLDAP

Armored-Dragon commented:
Help wanted as I have absolutely no idea how to get things to build with CMake.

Armored-Dragon committed:
Read from the admin panel for LDAP settings.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Housekeeping.
Removed outdated TODO/FIXMEs.
(Hopefully) fixed describe-settings.json regression.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Documentation.

Signed-off-by: armored-dragon <publicmail@armoreddragon.com>

Armored-Dragon committed:
Thats not LDAP.

AnotherFoxGuy committed:
:bug: Fixed missing OpenLDAP library

Armored-Dragon committed:
Merge pull request #95 from AnotherFoxGuy/fix-ldap-build

:bug: Fixed missing OpenLDAP library

JulianGro commented:
About OpenLDAP on Windows: While not open-source, the [Windows version](https://www.maxcrc.de/en/download-en/) does have a fully free license, so we could use it. It's a bit annoying that it isn't open-source, but personally I would be fine with having this.

Armored-Dragon commented:
> About OpenLDAP on Windows: While not open-source, the [Windows version](https://www.maxcrc.de/en/download-en/) does have a fully free license, so we could use it. It's a bit annoying that it isn't open-source, but personally I would be fine with having this.

I would greatly prefer we try and stick to open source only, but if there is not an alternative I won't stand in the way of it. In the future we can probably try and work something out to replace it with a open source version.

Armored-Dragon commented:
@ksuprynowicz I'm requesting a review from you since you have experience with LDAP systems. Let me know if I've done something silly!

JulianGro commented:
I wonder how hard it would be to compile this on Windows anyway. My understanding is that we only use the client library, while OpenLDAP is a suite also containing servers and utilities. It looks like we are building pretty much all of that. `make` also seems available on Windows through VCPKG.
It looks like the server can be disabled using `--disable-slapd`.
https://bin.linux.pizza/?0798efe40ebcb0d3#3UBwrzp2HhwD9WQRgB2f6NSUChUxU4GyeS4orkm8FDm5

ksuprynowicz commented:
@JulianGro There's an example of how to build it on Windows here:
https://github.com/python-ldap/python-ldap
It involves a pretty big patch file, but seems possible:
https://github.com/cgohlke/python-ldap-build/blob/main/openldap.diff

ksuprynowicz commented on libraries/networking/src/NodePermissions.h:
```diff
@@ -49,10 +49,10 @@ class NodePermissions {
     NodePermissionsKey getKey() const { return NodePermissionsKey(_id, _rankID); }
 
     // the _id member isn't authenticated/verified and _username is.
-    void setVerifiedUserName(QString userName) { _verifiedUserName = userName.toLower(); }
```
Maybe we should keep user names case-insensitive? On LDAP typically DNs are case-insensitive too.

Armored-Dragon replied:
I just revered the changes in the case sensitivity commit.

Armored-Dragon committed:
Revert case sensitivity.

Armored-Dragon committed:
Fixed duplicate semicolon.

ksuprynowicz commented on libraries/networking/src/LDAPAccount.cpp:
```diff
@@ -0,0 +1,187 @@
+        nullptr,
+        0,
+        &result
+    );
```
Does `result`  and other LDAP structures need to be manually disposed of to prevent memory leak?

Armored-Dragon replied:
They do. I was under the impression that it was handled automatically for some reason. I can't be sure what I was thinking three months ago, but this issue has since [been addressed](https://github.com/overte-org/overte/pull/1159/commits/c1ab5caabc92face710a2568ef7f15c3c81d61f0).

ksuprynowicz reviewed (commented):
Awesome!
Almost everything looks good, I only have a question related to possible memory leaks.

Armored-Dragon committed:
Fix duplicate connection denied packet.
https://github.com/overte-org/overte/pull/1159#discussion_r1929703881

Armored-Dragon committed:
Replace duplicate strings with a variable.
https://github.com/overte-org/overte/pull/1159#discussion_r1929702940

Armored-Dragon committed:
Reinserted a comment.

Armored-Dragon committed:
Replace commented code with a note.
https://github.com/overte-org/overte/pull/1159/files/622ab20842b82ddc95ba6b81269c9e91f4122c95#r1929707036

Armored-Dragon committed:
Fix typo.
https://github.com/overte-org/overte/pull/1159#discussion_r1929708879

Armored-Dragon commented on domain-server/src/DomainGatekeeper.cpp:
```diff
@@ -564,7 +588,32 @@ SharedNodePointer DomainGatekeeper::processAgentConnectRequest(const NodeConnect
+        else {
             sendConnectionDeniedPacket("You lack the required Directory Services permissions to connect to this domain.",
                 nodeConnection.senderSockAddr, DomainHandler::ConnectionRefusedReason::NotAuthorizedMetaverse);
         }
```
You are correct.

Armored-Dragon committed:
Login type agnostic function.
https://github.com/overte-org/overte/pull/1159#discussion_r1929706352

Armored-Dragon committed:
Dispose of LDAPConnections.
https://github.com/overte-org/overte/pull/1159#discussion_r1929724415

ksuprynowicz commented:
I was testing this PR this evening, and I cannot start the domain server. I get OpenSSL-related crash:
```
getrn 0x000055555589dd7b
OPENSSL_LH_retrieve 0x000055555589d80d
<unknown> 0x00007fffe8547a6c
<unknown> 0x00007fffe85472d6
<unknown> 0x00007fffe84569cc
<unknown> 0x00007fffe8456a3c
__pthread_once_slow 0x00007ffff58a1087
___pthread_once 0x00007ffff58a10f9
CRYPTO_THREAD_run_once 0x00005555558b2ad3
<unknown> 0x00007fffe8456cef
<unknown> 0x00007fffe84571bd
<unknown> 0x00007fffe8459a48
CRYPTO_get_ex_new_index 0x00007fffe8459c51
<unknown> 0x00007ffff7f1a148
<unknown> 0x00007ffff7f1e2a5
<unknown> 0x00007ffff7efc0a2
QSslConfiguration::defaultConfiguration() 0x00007ffff7ef271f
QNetworkRequest::sslConfiguration() const 0x00007ffff7e3b637
<unknown> 0x00007ffff7e9c74d
QNetworkAccessManager::createRequest(QNetworkAccessManager::Operation, QNetworkRequest const&, QIODevice*) 0x00007ffff7e2bbda
QNetworkAccessManager::put(QNetworkRequest const&, QIODevice*) 0x00007ffff7e2888a
QNetworkAccessManager::put(QNetworkRequest const&, QByteArray const&) 0x00007ffff7e288f5
AccountManager::sendRequest AccountManager.cpp:325
DomainServer::sendHeartbeatToMetaverse DomainServer.cpp:1643
DomainServer::sendHeartbeatToMetaverse DomainServer.h:112
DomainServer::DomainServer DomainServer.cpp:323
main main.cpp:49
__libc_start_call_main 0x00007ffff5833d68
__libc_start_main_impl 0x00007ffff5833e25
_start 0x000055555568b811
```

ksuprynowicz commented:
Can you rebase this PR on current master to see if that helps?

OfficialR3ido101 commented:
Tested on windows and it failed with message ```['C:\\Users\\olive/overte-files/vcpkg\\1bcc44a8\\vcpkg.exe', '--vcpkg-root', 'C:\\Users\\olive/overte-files/vcpkg\\1bcc44a8', 'install', '--triplet', 'x64-windows', 'hifi-client-deps']
Computing installation plan...
openldap is only supported on '!windows, (mingw & !x86)', which does not match x64-windows. This usually means that there are known build failures, or runtime problems, when building other platforms. To ignore this and attempt to build openldap anyway, rerun vcpkg with `--allow-unsupported`.
Traceback (most recent call last):
  File "D:\Code\armored\overte\prebuild.py", line 218, in <module>
    main()
    ~~~~^^
  File "D:\Code\armored\overte\prebuild.py", line 184, in main
    pm.setupDependencies(qt=qtInstallPath)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "D:\Code\armored\overte\hifi_vcpkg.py", line 271, in setupDependencies
    self.run(['install', '--triplet', self.getTripletWithBuildType(self.triplet), 'hifi-client-deps'])
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Code\armored\overte\hifi_vcpkg.py", line 229, in run
    hifi_utils.executeSubprocess(actualCommands, folder=self.path, env=self.buildEnv)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Code\armored\overte\hifi_utils.py", line 74, in executeSubprocess
    raise RuntimeError('Call to "{}" failed.\n\narguments:\n{}\n'.format(
    ...<2 lines>...
        ))
RuntimeError: Call to "C:\Users\olive/overte-files/vcpkg\1bcc44a8\vcpkg.exe" failed.

arguments:
--vcpkg-root C:\Users\olive/overte-files/vcpkg\1bcc44a8 install --triplet x64-windows hifi-client-deps

CMake Error at CMakeLists.txt:214 (message):
  prebuild.py failed with error 1


-- Configuring incomplete, errors occurred!``` This is when trying to get cmake to gather packages.
